### PR TITLE
fuse: support the new macFUSE mount protocol

### LINKED ISSUE #379: go-fuse currently not support macfuse 4.0.0
[osxfuse](https://osxfuse.github.io/) changed its name to macfuse since version [4.0.0](https://github.com/osxfuse/osxfuse/releases/tag/macfuse-4.0.0) released. Below paragraph is copied from [changelog](https://github.com/osxfuse/osxfuse/releases/tag/macfuse-4.0.0):
- Drop support for deprecated "old style" mount process. Opening a /dev/macfuse device and passing the file descriptor to mount_macfuse is no longer supported ("old style"). mount_macfuse passes the file descriptor to the calling library over a socket ("new style", introduced in version 3.3.0).

In [changlog](https://github.com/osxfuse/osxfuse/releases/tag/osxfuse-3.3.0) of version 3.3.0, it states this:
- Implement new style mount process

  When mounting a FUSE volume on Linux, fusermount opens the FUSE device and passes the file descriptor back to libfuse.

  However, on OS X libosxfuse used to be responsible for locating a free FUSE device, opening it, and then passing the file descriptor to mount_osxfuse. For this to work libosxfuse needs to be aware of low level implementation details like the number of registered FUSE devices.

  In order to decouple libosxfuse from the FUSE kernel extension the mount process has been revised to match FUSE on Linux. mount_osxfuse will locate a free FUSE device, open it, and then pass the file descriptor back to libosxfuse.

go-fuse currently seems to depend on this "old style".

### PR DESCRIPTION
**Pull request by midchildan**
Closes #379.

~This is still a work in progress.~

# Changes

- adds support for the new style mount process detailed in https://github.com/osxfuse/fuse/commit/8afa4eabdddb7fc9e810f43666462c4d2451925f
- drops support for the old style mount process
- support paths and env vars renamed in the osxfuse -> macFUSE rebranding changes

### CODE REVIEW

**Comment by hanwen** [comment]
> This is still a work in progress.

what does this mean?

Note that I'll need a CLA.

**Comment by midchildan** [comment]
It didn't quite work when I created the PR. I've fixed some stuff since then and the hello example now works with osxfuse 3.x using the new style mount process.

I've also ran the tests and it seems to fail with zipfs, unionfs, and fuse/test. They also fail in master, so I guess it's out of the scope for this PR.

I haven't yet checked if this PR works with macFUSE 4.x. I'll leave that for tomorrow since it's getting late in my timezone ;-)

**Comment by midchildan** [comment]
@googlebot I signed it!

**Comment by midchildan** [comment]
I've looked into the tests a little further and it'll likely involve bumping the kernel interface version to get it fixed. I'll see if existing applications would work as it did before with just the current changes.

**Comment by midchildan** [comment]
This PR is now ready for review. I've confirmed that this PR changes work with gocryptfs, on both osxfuse 3.11.2 and macFUSE 4.0.4. I was able to read, write, modify, rename, and delete files and directories on a gocryptfs filesystem.

As a side note, this change drops support for the old style mount process completely. I don't think this would be a problem, because the new style mount process has been supported since [osxfuse 3.3.0][1], which was released back in 2016. This should be old enough that it wouldn't run on any currently supported version of macOS.

[1]: https://github.com/osxfuse/osxfuse/releases/tag/osxfuse-3.3.0

**Comment by llucps** [comment]
Thank you @midchildan !

**Comment by wmchris** [comment]
As this is mandatory for OSX users with BigSur and M1 - any chance to proritize this pull request?

File: fuse/mount_darwin.go [hunk 1]
```diff
@@ -9,89 +9,124 @@ import (
...
-		err := cmd.Wait()
-		if err != nil {
-			err = fmt.Errorf("mount_osxfusefs failed: %v. Stderr: %s, Stdout: %s", err, errOut.String(), out.String())
+		if err := cmd.Wait(); err != nil {
+			err = fmt.Errorf("mount_osxfusefs failed: %v. Stderr: %s, Stdout: %s",
+				err, errOut.String(), out.String())
 		}
 
 		ready <- err
 		close(ready)
```
Comment (hanwen) [inline fuse/mount_darwin.go:77]: is there still a reason to use a goroutine here? Could this just wait and close(ready) inline like the Linux support does?

If no, document why.

**Review by hanwen — COMMENTED** [review]
I actually prefer doing reviews on Gerrithub (see CONTRIBUTING)

File: fuse/mount_darwin.go (hunk 1 above)
Comment (midchildan) [reply fuse/mount_darwin.go:78]: It just blocked forever when I removed the goroutine. I added a comment explaining this.

**Comment by midchildan** [comment]
I've also pushed the changes to gerrithub https://review.gerrithub.io/c/hanwen/go-fuse/+/508417

**Comment by hanwen** [comment]
> I've also pushed the changes to gerrithub https://review.gerrithub.io/c/hanwen/go-fuse/+/508417

Oh, saw it too late. I merged it already as https://github.com/hanwen/go-fuse/commit/09a3c381714cf1011fb2d08885f29896cd496a0c

**Comment by grrywlsn** [comment]
Thanks for the work on this! Out of curiosity, what is the release strategy for `go-fuse` - will this change make it to 2.0.4?

**Comment by hanwen** [comment]
there is no documented strategy. Would you like me tag a release?

**Comment by grrywlsn** [comment]
Yes please @hanwen - M1 Macs and Intel Macs using v4 of macFUSE depend on one of the more recent merges as a fix, so having that in a tagged version would be great.

**Comment by llucps** [comment]
I guess it's me, but although I was able to install go and successfully compile gocryptfs following the documentation. I've been able to make it work.
I uninstalled osxfuse v3, restart, install the new macFuse 4 and restart again when I try to run:

```
./gocryptfs xxx xxxx

Complied version -> gocryptfs v2.0-beta2-9-gbed6010 without_openssl; go-fuse v2.0.4-0.20201104153454-be8e5f4a85fd; 2021-01-31 go1.15.6 darwin/amd64
```

I get:
```
fs.Mount failed: fork/exec /Library/Filesystems/osxfusefs.fs/Support/load_osxfusefs: no such file or directory
Maybe you should run: /Library/Filesystems/osxfuse.fs/Contents/Resources/load_osxfuse
```

It seems is still looking for osxfuse v3.. I guess I'm doing something wrong.

Any help would be appreciated.
Thanks.

**Comment by midchildan** [comment]
This PR removed all references to `load_osxfusefs`. So looking at your error message, you likely haven't compiled go-fuse with the latest changes included.

**Comment by midchildan** [comment]
You're using go-fuse be8e5f4a85fd, so it doesn't include this change. [The latest gocryptfs uses the latest commit of go-fuse](https://github.com/rfjakob/gocryptfs/blob/3ab1bcf1c5aacafe4339564ff0a09c349d3201a0/go.mod#L6), so it should work with macFUSE 4.x out of the box.

**Comment by llucps** [comment]
Thanks @midchildan.

Yes you're right, as soon as I updated the packages then it compiled successfully with the latest chages.

Although now I have another problem which is I installed macFuse 4.0.5 and according to https://github.com/osxfuse/osxfuse/issues/741 there should be a folder in /Library/StagedExtensions/Library/Filesystems/macfuse.fs/Contents/Extensions/11/macfuse.kext which I don't have in my system despite macFuse being installed..

As soon as I try to run:
`./gocryptfs -fusedebug -q -ko local,allow_other,default_permissions,auto_xattr,no_readahead /XXX /XXX/
`
I get:

`08:14:29.652369 rx 2: INIT n0 {7.19 Ra 0x200000 VOL_RENAME,XTIMES,EXPLICIT_INVAL_DATA,CASE_INSENSITIVE,ATOMIC_O_TRUNC,0xc000000}
08:14:29.652503 tx 2:     OK, {7.8 Ra 0x200000  0/0 Wr 0x20000 Tg 0x0}
08:14:29.652739 rx 3: STATFS n1
08:14:29.652814 tx 3:     OK, {blocks (109271,92870)/239172 files 2448172113/2449125360 bs1048576 nl0 frs1048576}
08:14:29.652861 rx 2: STATFS n1
08:14:29.652920 tx 2:     OK, {blocks (109271,92870)/239172 files 2448172113/2449125360 bs1048576 nl0 frs1048576}
08:14:29.653264 rx 3: STATFS n1
08:14:29.653299 rx 2: STATFS n1
08:14:29.653333 tx 2:     OK, {blocks (109271,92870)/239172 files 2448172113/2449125360 bs1048576 nl0 frs1048576}
08:14:29.653333 tx 3:     OK, {blocks (109271,92870)/239172 files 2448172113/2449125360 bs1048576 nl0 frs1048576}
08:14:29.653454 rx 2: STATFS n1
08:14:29.653528 tx 2:     OK, {blocks (109271,92870)/239172 files 2448172113/2449125360 bs1048576 nl0 frs1048576}
08:14:29.653926 rx 3: STATFS n1
08:14:29.653971 tx 3:     OK, {blocks (109271,92870)/239172 files 2448172113/2449125360 bs1048576 nl0 frs1048576}
08:14:29.654523 rx 2: GETATTR n1
08:14:29.654759 tx 2:     OK, {tA=1s {M040755 SZ=736 L=23 501:20 0 0:0 A 1607963224.871232 M 1607963224.795708 C 1607963224.795708}}
08:14:29.654874 rx 3: LOOKUP n1 [".go-fuse-epoll-hack"] 20b
08:14:29.654920 tx 3:     OK, {n18446744073709551615 g0 tE=0s tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
08:14:29.654998 rx 2: GETATTR n18446744073709551615
08:14:29.655034 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
08:14:29.655156 rx 3: GETATTR n18446744073709551615
08:14:29.655193 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
08:14:29.655344 rx 2: GETATTR n18446744073709551615
08:14:29.655380 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
08:14:29.655448 rx 3: GETATTR n18446744073709551615
08:14:29.655519 tx 3:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
08:14:29.655594 rx 2: GETATTR n18446744073709551615
08:14:29.655630 tx 2:     OK, {tA=0s {M0100644 SZ=0 L=1 0:0 0 0:18446744073709551615 A 0.000000 M 0.000000 C 0.000000}}
fs.Mount failed: permission denied`

I'll try to debug a bit more.

Thanks.

**Comment by llucps** [comment]
Ok finally manage to make it work.

For some reason the `fs.Mount Failed: permission denied` was because of the option `default_permissions`. I removed from the command and it works now.

Thanks.

**Comment by grrywlsn** [comment]
@hanwen are you happy to tag a release now that this seems to be solved?

**Pull request closed without merging**